extension/injections/shared_factory.cpp

```diff
@@ -12,6 +12,25 @@
 
 namespace di = boost::di;
 
+//<<rebind only if expected and current type are same>>
+template <class TExpected, class T, class = void>
```

Comment:
Did you consider not using enable_if here and just use template specialization instead?

```cpp
template <class TExpected, class T>
struct injector_rebinder {
  template <class TInjector>
  auto rebind(TInjector& injector) {
    return std::move(injector);
  }
};

template <class T>
struct injector_rebinder<T, T> {
  template <class TInjector>
  auto rebind(TInjector& injector) {
    return di::make_injector(std::move(injector), di::bind<T>().in(di::unique)[di::override]);
  }
};
```

Reply:
@krzysztof-jusiak thank you, that is exactly what I was looking for :)